### ISSUE TITLE
Add support for asynchronous reward functions in GRPOTrainer (and maybe other trainers)

### ISSUE THREAD

**Issue by ideechy** [huggingface/trl#3426]
### Feature request

Detect when a reward_func is a coroutine, launch all such functions with asyncio.gather, and insert their results into the existing rewards_per_func tensor exactly as for synchronous rewards.
This enables users to mix GPU‑bound nn.Module rewards with latency‑bound API calls in the same training loop without changing the public API.

[current implementation](https://github.com/huggingface/trl/blob/4da4dc9117e04d3902eb5308f5bbf02faa8e9bc8/trl/trainer/grpo_trainer.py#L1141)

### Motivation

When reward functions call external services the current sequential evaluation in GRPOTrainer causes wall‑clock time to scale linearly with #reward functions × batch size. This quickly becomes the main bottleneck.

### Your contribution

The following is an sketch of my implementation
```
# in trl/trainer/grpo_trainer.py
import inspect, asyncio

rewards_per_func = torch.zeros(len(prompts), len(self.reward_funcs), device=device)

# --- split rewards ---
sync_funcs, async_funcs = [], []
for idx, (f, proc_cls) in enumerate(zip(self.reward_funcs, self.reward_processing_classes)):
    (async_funcs if inspect.iscoroutinefunction(f) else sync_funcs).append((idx, f, proc_cls))

# --- run synchronous rewards (existing logic) ---

# --- run asynchronous rewards ---
if async_funcs:
    async def _eval_async():
        tasks = [f(prompts=prompts, completions=completions, **reward_kwargs)
                 for _, f, _ in async_funcs]
        return await asyncio.gather(*tasks)

    results = asyncio.run(_eval_async())
    for (idx, _, _), res in zip(async_funcs, results):
        res = [r if r is not None else torch.nan for r in res]
        rewards_per_func[:, idx] = torch.tensor(res, dtype=torch.float32, device=device)
```
I've experimented with this change internally and see a significant speed up for training. In our setting, we use 2-6 rewards returned by calling Gemini/ChatGPT, sometime each call has ~500ms latency. Before this change we need to make each call synchronous. Attched are flamegraph comparison of the training time split before and after change.

Before: reward evaluation taking more than half of the time of each step

![Image](https://github.com/user-attachments/assets/9a7744cd-e3f0-4926-b7fa-2a183d8b8335)

After: generation step is now the main part of a training step

![Image](https://github.com/user-attachments/assets/84c5afcb-14e5-43e8-8a55-058fa7012a19)

I'd be happy to create a PR for that.

**Comment by HorHang** [comment]
[+ 1]

Any update on this feature? For some specific use cases, the reward function computation is take long time. So, GPU become idle at that time as shown in sharp decreasing of GPU utilization. The problem and solution is implemented and point out at verl https://github.com/volcengine/verl/issues/1584#issuecomment-2918528258.

Hope to see this feature soon!!!

![Image](https://github.com/user-attachments/assets/c7a252d7-561a-422f-b9a6-4bd114bd6128)

**Comment by skerit** [comment]
@ideechy Do you have your async implementation public somewhere?

**Comment by ideechy** [comment]
yes https://github.com/huggingface/trl/compare/main...aideate:trl:haoyu/async_reward

**Comment by d-falck** [comment]
Also keen to see this

**Comment by qgallouedec** [comment]
see #4130

**Comment by pramodith** [comment]
@qgallouedec @albertvillanova I can open up a PR for this if no one else is actively working on it.

**Comment by qgallouedec** [comment]
Hey @ideechy, sorry for the long delay here, just catching up on the backlog. Good news: async reward functions are now supported in both GRPOTrainer (PR #4567) and RLOOTrainer (PR #4718); the implementation matches what you sketched (split sync/async by `inspect.iscoroutinefunction`, then `asyncio.gather` the async ones). You can see the GRPO path here: https://github.com/huggingface/trl/blob/78776953c1c49253787ea04c23e9390ffd14fe02/trl/trainer/grpo_trainer.py#L1213-L1265. Closing as fixed. Thanks for the detailed report and flamegraphs that motivated the design.
